### PR TITLE
Model checkpointing

### PR DESCRIPTION
**Pull request by dhpitt**
This PR implements two model checkpointing callbacks. One simply saves a model every N epochs, and the other monitors a metric from `eval_losses` to determine when to save a model.

### CODE REVIEW

File: neuralop/training/trainer.py [hunk 1]
```diff
@@ -69,6 +70,8 @@ def __init__(self, *,
                  use_distributed=use_distributed, 
                  verbose=verbose)
 
+        if load_from_checkpoint:
+            self.model = torch.load()
```
Comment (JeanKossaifi) [inline neuralop/training/trainer.py:74]: We should be updating the state dict of the model here not overwriting it

File: neuralop/training/callbacks.py [hunk 2]
```diff
@@ -346,4 +343,84 @@ def on_before_loss(self, out):
...
+            ckpt_path = self.ckpt_dir / f"{self.state_dict['epoch']}"
+            torch.save(self.state_dict['model'], ckpt_path)
+        
+
+class MonitorMetricCheckpointCallback(ModelCheckpointCallback):
+    """
+    Implements model checkpointing with the addition of monitoring a metric
+    """
+
+    def __init__(self, monitor: str, ckpt_dir: str = './checkpoints'):
```
Comment (JeanKossaifi) [inline neuralop/training/callbacks.py:388]: can we weite explicitly checkpoint_path?

Comment (JeanKossaifi) [reply neuralop/training/callbacks.py:388]: And monitor could be e.g. loss_key

Comment (JeanKossaifi) [reply neuralop/training/callbacks.py:388]: (same for above method for regular checkpoint)

**Review by JeanKossaifi — COMMENTED** [review]
Looks good overall @dhpitt !
